File: webapp-vue/src/lib/rng/seededRandom.ts
// Mirror of backend SeededRandom (Kotlin is the source of truth). Byte-identical streams are a
// contract verified against shared/rng/golden-vectors.json — see seededRandom.spec.ts.
//
// nextUint32 follows Blackman/Vigna's reference verbatim — https://prng.di.unimi.it/xoshiro128starstar.c
// — i.e. v1.1, whose scrambler reads s1. The superseded v1.0 reads s0 and yields a different stream.
//
// Only operations whose result is mandated bit-for-bit by both ECMAScript and the JVM are used:
// ToInt32/ToUint32 bit ops, Math.imul, and IEEE754 add/multiply/divide by powers of two.
// Math.sin/cos/log/exp/pow/atan/cbrt are implementation-approximated and DIVERGE between engines
// (measured: V8 vs JavaScriptCore differ by 1 ulp) — they must never enter this file.

const FNV_OFFSET_BASIS = 0x811c9dc5
const FNV_PRIME = 0x01000193
const GOLDEN_GAMMA_32 = 0x9e3779b9
const TWO_POW_32 = 0x100000000
const TWO_POW_26 = 67108864
const TWO_POW_53 = 9007199254740992

const utf8 = new TextEncoder()

const rotl = (x: number, k: number): number => (x << k) | (x >>> (32 - k)) | 0

/** FNV-1a over the UTF-8 bytes — NOT over UTF-16 code units, which would diverge for 'hütte'. */
function hashSeedString(seed: string): number {
  let h = FNV_OFFSET_BASIS | 0
  for (const byte of utf8.encode(seed)) {
    h ^= byte
    h = Math.imul(h, FNV_PRIME)
  }
  return h | 0
}

export class SeededRandom {
  private s0: number
  private s1: number
  private s2: number
  private s3: number

  private constructor(s0: number, s1: number, s2: number, s3: number) {
    this.s0 = s0
    this.s1 = s1
    this.s2 = s2
    this.s3 = s3
  }

  /** A string seed is hashed to 32 bits; a number seed is truncated to its low 32 bits. */
  static fromSeed(seed: number | string): SeededRandom {
    const seed32 = typeof seed === 'string' ? hashSeedString(seed) : seed | 0
    let a = seed32
    const state = new Int32Array(4)
    for (let i = 0; i < 4; i++) {
      a = (a + GOLDEN_GAMMA_32) | 0
      let t = a ^ (a >>> 16)
      t = Math.imul(t, 0x21f0aaad)
      t = t ^ (t >>> 15)
      t = Math.imul(t, 0x735a2d97)
      state[i] = t ^ (t >>> 15)
    }
    const [s0 = 0, s1 = 0, s2 = 0, s3 = 0] = state
    // xoshiro cannot escape the all-zero state; splitmix32 can only produce it for one seed.
    return (s0 | s1 | s2 | s3) === 0
      ? new SeededRandom(1, s1, s2, s3)
      : new SeededRandom(s0, s1, s2, s3)
  }

  /** xoshiro128** — the raw stream. Returns an unsigned 32-bit value as a safe integer. */
  nextUint32(): number {
    const result = Math.imul(rotl(Math.imul(this.s1, 5), 7), 9)
    const t = (this.s1 << 9) | 0
    this.s2 ^= this.s0
    this.s3 ^= this.s1
    this.s1 ^= this.s2
    this.s0 ^= this.s3
    this.s2 ^= t
    this.s3 = rotl(this.s3, 11)
    return result >>> 0
  }

  /** Unbiased via rejection — plain `% span` would over-represent the low residues. */
  private nextBounded(span: number): number {
    const threshold = TWO_POW_32 % span
    let r = this.nextUint32()
    while (r < threshold) r = this.nextUint32()
    return r % span
  }

  nextInt(boundExclusive: number): number {
    if (!Number.isInteger(boundExclusive) || boundExclusive <= 0) {
      throw new RangeError(`bound must be a positive integer, was ${boundExclusive}`)
    }
    return this.nextBounded(boundExclusive)
  }

  nextIntBetween(min: number, maxInclusive: number): number {
    if (maxInclusive < min) {
      throw new RangeError(`max must not be below min, was ${min}..${maxInclusive}`)
    }
    return min + this.nextBounded(maxInclusive - min + 1)
  }

  /** 53 bits from two words, scaled by an exact power of two, so the value is bit-identical. */
  nextDouble(): number {
    const hi = this.nextUint32() >>> 5
    const lo = this.nextUint32() >>> 6
    return (hi * TWO_POW_26 + lo) / TWO_POW_53
  }

  /** The most significant bit. Switching to `& 1` would need the vectors regenerated on both sides. */
  nextBoolean(): boolean {
    return this.nextUint32() >>> 31 !== 0
  }

  // Fisher-Yates, descending, drawing nextInt(i + 1). Drawing nextInt(length) instead is genuinely
  // biased; running ascending is unbiased but yields a different permutation. Both mistakes
  // reproduce identically on the JVM, so no parity test catches them — the direction is pinned only
  // by the golden vectors.
  shuffled<T>(items: readonly T[]): T[] {
    const out = items.slice()
    for (let i = out.length - 1; i > 0; i--) {
      const j = this.nextInt(i + 1)
      // Both indices are in bounds, so the reads cannot be undefined.
      const swapped = out[i] as T
      out[i] = out[j] as T
      out[j] = swapped
    }
    return out
  }

  pick<T>(items: readonly T[]): T {
    if (items.length === 0) throw new RangeError('cannot pick from an empty list')
    return items[this.nextInt(items.length)] as T
  }

  /** Weights are summed left to right so the JVM performs the identical sequence of additions. */
  weightedPick<T>(items: readonly T[], weights: readonly number[]): T {
    if (items.length === 0 || items.length !== weights.length) {
      throw new RangeError(`items and weights must be non-empty and equal in length`)
    }
    let total = 0
    for (const w of weights) {
      if (!(w >= 0)) throw new RangeError(`weights must be non-negative, was ${w}`)
      total += w
    }
    if (total <= 0) throw new RangeError('weights must not sum to zero')
    const target = this.nextDouble() * total
    let cumulative = 0
    for (let i = 0; i < items.length; i++) {
      cumulative += weights[i] as number
      if (target < cumulative) return items[i] as T
    }
    return items[items.length - 1] as T
  }

  /** Without replacement. Equivalent to taking the first k of a full shuffle. */
  sample<T>(items: readonly T[], k: number): T[] {
    if (k < 0 || k > items.length) {
      throw new RangeError(`k must be within 0..${items.length}, was ${k}`)
    }
    return this.shuffled(items).slice(0, k)
  }
}

File: core/src/main/kotlin/org/unividuell/countdown/core/rng/SeededRandom.kt
package org.unividuell.countdown.core.rng

/**
 * xoshiro128** with splitmix32 seed expansion. Reproducible by design: a round can be re-derived from
 * its seed at any time, so nothing generated has to be persisted. No platform RNG offers that —
 * `kotlin.random.Random` disclaims cross-version stability, and `L32X64MixRandom` changed its
 * per-seed stream in an OpenJDK patch release (JDK-8282551).
 *
 * The stream is also reproducible in a browser runtime, pinned by `shared/rng/golden-vectors.json`
 * against a test-scope reference implementation in
 * `webapp-vue/src/lib/rng/__tests__/seededRandom.reference.ts`. That property is verified but
 * deliberately unused: rounds are server-authoritative, and a seed that reaches the client reveals
 * every future draw.
 *
 * [nextUint32] follows Blackman/Vigna's reference verbatim — https://prng.di.unimi.it/xoshiro128starstar.c
 * — i.e. **v1.1, whose scrambler reads `s1`**. The superseded v1.0 reads `s0` and yields a different
 * stream; Apache Commons RNG still ships that older variant, so "aligning with the library" would
 * break the contract rather than fix it.
 *
 * Only operations mandated bit-for-bit by both the JVM and ECMAScript are used: 32-bit two's
 * complement arithmetic and IEEE754 add/multiply/divide by powers of two. `Math.log`, `sin`, `cos`,
 * `exp`, `pow`, `atan` and `cbrt` are implementation-approximated on both platforms and measurably
 * disagree between JS engines (V8 vs JavaScriptCore differ by 1 ulp), so they must never appear here.
 *
 * The `rng` module owns no tables and holds no state beyond an instance's four words, so it needs
 * neither a schema nor a Flyway migration. It is the exposed API of the module: construct one per
 * seeded unit of work (a round, a puzzle) and let it go — never share an instance across requests,
 * since every draw mutates the state and would desync a concurrent replay.
 *
 * See `docs/superpowers/specs/2026-08-02-cross-runtime-rng-design.md`.
 */
class SeededRandom private constructor(
    private var s0: Int,
    private var s1: Int,
    private var s2: Int,
    private var s3: Int,
) {

    companion object {
        private const val FNV_OFFSET_BASIS = -0x7ee3623b // 0x811c9dc5
        private const val FNV_PRIME = 0x01000193
        private const val GOLDEN_GAMMA_32 = -0x61c88647 // 0x9e3779b9
        private const val TWO_POW_32 = 0x1_0000_0000L
        private const val TWO_POW_26 = 67108864.0
        private const val TWO_POW_53 = 9007199254740992.0

        fun fromSeed(seed: Int): SeededRandom {
            var a = seed
            val state = IntArray(4)
            for (i in 0 until 4) {
                a += GOLDEN_GAMMA_32
                var t = a xor (a ushr 16)
                t *= 0x21f0aaad
                t = t xor (t ushr 15)
                t *= 0x735a2d97
                state[i] = t xor (t ushr 15)
            }
            // xoshiro cannot escape the all-zero state; splitmix32 can only produce it for one seed.
            if (state[0] or state[1] or state[2] or state[3] == 0) state[0] = 1
            return SeededRandom(state[0], state[1], state[2], state[3])
        }

        /** Hashed over the UTF-8 bytes — hashing UTF-16 code units would diverge for "hütte". */
        fun fromSeed(seed: String): SeededRandom {
            var h = FNV_OFFSET_BASIS
            for (byte in seed.toByteArray(Charsets.UTF_8)) {
                h = h xor (byte.toInt() and 0xff)
                h *= FNV_PRIME
            }
            return fromSeed(h)
        }

        private fun rotl(x: Int, k: Int): Int = (x shl k) or (x ushr (32 - k))
    }

    /** The raw xoshiro128** stream, widened so the value is unsigned and comparisons are safe. */
    fun nextUint32(): Long {
        val result = rotl(s1 * 5, 7) * 9
        val t = s1 shl 9
        s2 = s2 xor s0
        s3 = s3 xor s1
        s1 = s1 xor s2
        s0 = s0 xor s3
        s2 = s2 xor t
        s3 = rotl(s3, 11)
        return result.toUInt().toLong()
    }

    /** Unbiased via rejection — a plain `% span` would over-represent the low residues. */
    private fun nextBounded(span: Long): Long {
        val threshold = TWO_POW_32 % span
        var r = nextUint32()
        while (r < threshold) r = nextUint32()
        return r % span
    }

    fun nextInt(boundExclusive: Int): Int {
        require(boundExclusive > 0) { "bound must be positive, was $boundExclusive" }
        return nextBounded(boundExclusive.toLong()).toInt()
    }

    fun nextIntBetween(min: Int, maxInclusive: Int): Int {
        require(maxInclusive >= min) { "max must not be below min, was $min..$maxInclusive" }
        val span = maxInclusive.toLong() - min.toLong() + 1L
        return (min + nextBounded(span)).toInt()
    }

    /** 53 bits from two words, scaled by an exact power of two, so the result is bit-identical. */
    fun nextDouble(): Double {
        val hi = nextUint32() ushr 5
        val lo = nextUint32() ushr 6
        return (hi * TWO_POW_26 + lo) / TWO_POW_53
    }

    /**
     * The most significant bit — the conservative choice, and the one `java.util.Random.next(1)`
     * makes too. Switching to `and 1` would need the vectors regenerated on both sides.
     */
    fun nextBoolean(): Boolean = (nextUint32() ushr 31) != 0L

    /**
     * Fisher-Yates, descending, drawing `nextInt(i + 1)`.
     *
     * Drawing `nextInt(size)` instead is genuinely biased. Running ascending is *not* biased but
     * produces a different permutation for the same seed — and, like the biased variant, it
     * reproduces identically on both platforms, so no parity test would catch either mistake. The
     * direction is pinned only by the golden vectors. `Collections.shuffle`/`MutableList.shuffle`
     * must not be used: their draw order is tied to `java.util.Random`, which has no browser
     * counterpart.
     */
    fun <T> shuffled(items: List<T>): List<T> {
        val out = items.toMutableList()
        for (i in out.size - 1 downTo 1) {
            val j = nextInt(i + 1)
            val swapped = out[i]
            out[i] = out[j]
            out[j] = swapped
        }
        return out
    }

    fun <T> pick(items: List<T>): T {
        require(items.isNotEmpty()) { "cannot pick from an empty list" }
        return items[nextInt(items.size)]
    }

    /** Weights are summed left to right so the browser performs the identical additions. */
    fun <T> weightedPick(items: List<T>, weights: List<Double>): T {
        require(items.isNotEmpty() && items.size == weights.size) {
            "items and weights must be non-empty and equal in length"
        }
        var total = 0.0
        for (w in weights) {
            require(w >= 0.0) { "weights must be non-negative, was $w" }
            total += w
        }
        require(total > 0.0) { "weights must not sum to zero" }
        val target = nextDouble() * total
        var cumulative = 0.0
        for (i in items.indices) {
            cumulative += weights[i]
            if (target < cumulative) return items[i]
        }
        return items[items.size - 1]
    }

    /** Without replacement. Equivalent to taking the first k of a full shuffle. */
    fun <T> sample(items: List<T>, k: Int): List<T> {
        require(k in 0..items.size) { "k must be within 0..${items.size}, was $k" }
        return shuffled(items).take(k)
    }
}
